feat: support for tryit buttons that send snippets to live vscode editor.

lovettchris opened the pull request:
This is a work in progress..

https://user-images.githubusercontent.com/18707114/149084867-78d83dd9-7a75-43e1-92c5-19824b76d10a.mp4

.

gebner commented:
- The reference manual opens in the browser, not in vscode.
- The "Example" link opens a file that doesn't compile.
- The TPIL4 font is light gray (on white background) and the copy button doesn't open the file in vscode (but I guess that's because the documentation hasn't been updated yet):
![grayfontcolor](https://user-images.githubusercontent.com/313929/151341686-8ae437ff-e5cd-44f4-ab89-d5dca31f64e3.png)
- The back button doesn't work in TPIL4.

lovettchris commented:
I added missing instructions on how to test it locally in the overview above.

1. I fixed the reference manual link to run inside vscode.
2. The "Example" link is a placeholder for a new "lean4-samples" repo that Leo is creating.
3. I'm working on fixing how the theme changes are synchronized... light theme isn't working yet...

gebner commented:
It's a bit awkward that we have to have such a tight integration with the mdbook tool.  Users also want to use their custom documentation generators; e.g. sphinx is fairly popular.  But we can add support for that when the time comes.

I was actually really surprised to see these kind of theming issues here.  I didn't recall any such issues from the Lean 3 extension.  It turns out that sphinx themes set both `background` and `color` properties, which override the properties set by vscode.

Let's wait with merging until we can switch to the official links for TPIL4 and the reference manual.

The back button still doesn't work btw.  I think you just need to increase the z-index.

lovettchris commented:
> I was actually really surprised to see these kind of theming issues here. I didn't recall any such issues from the Lean 3 extension. It turns out that sphinx themes set both `background` and `color` properties, which override the properties set by vscode.

I'm confused, what theming issues?  Were you expecting the mdbook themes to pickup the vscode themes automatically or something?  Do you know how that was supposed to work?  I couldn't see anything in the mdbook styles that would do that.  And the mdbook styles default to "navy" but "coal" looks better in VSCode... Hence the translation thunk that I'm adding with the vscode themes.ts overrides.

lovettchris commented:
> The back button still doesn't work btw. I think you just need to increase the z-index.

Fixed.

gebner commented:
> I'm confused, what theming issues?

Just the fact the we have to set the theme at all, and that the colors are awful otherwise.  We didn't have to do that in Lean 3 (because the sphinx styles set explicit background and foreground colors).

> I also don't know a more general way of doing acquireVsCodeApi. Most "webview" hosting apps provide a callback api like this that you can call "sendMessage" on

I'm not talking about extensibility on this side, we're not going to use the `themes.js` code outside of vscode.  I am talking about the `set_theme` function, which is only implemented in (our special) fork of mdbook.

gebner commented:
In any case, we can iterate on that extensibility later.

I'm just waiting for https://leanprover.github.io/lean4/doc/ and https://leanprover.github.io/theorem_proving_in_lean4/ to be upgraded to the new mdbook version before merging.

lovettchris commented:
BTW: I just tried Lean3 and the themes are wrong, so I'm guessing lean3 never tried to integrate with vscode themes like I have done in this PR:

![image](https://user-images.githubusercontent.com/18707114/152059655-d1c8a1ed-a4df-49f9-ad19-1fce7c506245.png)

lovettchris commented:
@gebner this PR is now ready to merge, Leo just published a new version of the TPIL book built using the new version of mdbook so it has the correct book.js that works nicely inside vs code now:

![image](https://user-images.githubusercontent.com/18707114/153663011-f658f8fa-ab0d-433a-b1bc-8806c4ba164d.png)

lovettchris commented on vscode-lean4/src/docview.ts:
```diff
@@ -163,10 +233,16 @@ export class DocViewProvider implements Disposable {
+                'Theorem Proving in Lean': mkCommandUri('lean4.docView.open', 'http://lovettsoftware.com/theorem_proving_in_lean4/introduction.html'),
+                'Reference Manual': mkCommandUri('lean4.docView.open', 'https://leanprover.github.io/lean4/doc/'),
+                'Abbreviations cheat sheet': mkCommandUri('lean4.docView.showAllAbbreviations'),
+                'Example': mkCommandUri('lean4.openExample', 'https://raw.githubusercontent.com/leanprover/lean4/master/doc/examples/compiler/test.lean'),
```
Fixed, it now points to our new repo https://github.com/leanprover/lean4-samples/tree/main/HelloWorld.  Eventually I will change up how the "Example" link is generated by instead jumping to an index provided on https://github.com/leanprover/lean4-samples that contains some sort of special annotation so we can find and create the appropriate `mkCommandUri('lean4.openExample',...)` links.
